<a href="https://colab.research.google.com/github/Datta0/notebooks/blob/main/nb/Qwen3_(4B)-GRPO.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

**NEW** Unsloth now supports training the new **gpt-oss** model from OpenAI! You can start finetune gpt-oss for free with our **[Colab notebook](https://x.com/UnslothAI/status/1953896997867729075)**!

Unsloth now supports Text-to-Speech (TTS) models. Read our [guide here](https://docs.unsloth.ai/basics/text-to-speech-tts-fine-tuning).

Read our **[Gemma 3N Guide](https://docs.unsloth.ai/basics/gemma-3n-how-to-run-and-fine-tune)** and check out our new **[Dynamic 2.0](https://docs.unsloth.ai/basics/unsloth-dynamic-2.0-ggufs)** quants which outperforms other quantization methods!

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [1]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth vllm
else:
    # [NOTE] Do the below ONLY in Colab! Use [[pip install unsloth vllm]]
    !pip install --no-deps unsloth vllm==0.8.5.post1

In [2]:
#@title Colab Extra Install { display-mode: "form" }
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth vllm
else:
    !pip install --no-deps unsloth vllm==0.8.5.post1
    # [NOTE] Do the below ONLY in Colab! Use [[pip install unsloth vllm]]
    # Skip restarting message in Colab
    import sys, re, requests; modules = list(sys.modules.keys())
    for x in modules: sys.modules.pop(x) if "PIL" in x or "google" in x else None
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf "datasets>=3.4.1,<4.0.0" "huggingface_hub>=0.34.0" hf_transfer

    # vLLM requirements - vLLM breaks Colab due to reinstalling numpy
    f = requests.get("https://raw.githubusercontent.com/vllm-project/vllm/refs/heads/main/requirements/common.txt").content
    with open("vllm_requirements.txt", "wb") as file:
        file.write(re.sub(rb"(transformers|numpy|xformers)[^\n]{1,}\n", b"", f))
    !pip install -r vllm_requirements.txt

### Unsloth

Goal: To convert `Qwen3-4B-Base` into a reasoning model via GRPO by using OpenR1's Math dataset.

We first pre fine-tune the model to make GRPO skip trying to match formatting - this speeds GRPO up.

In [3]:
import os
os.environ['UNSLOTH_VLLM_STANDBY'] = '1'
os.environ["UNSLOTH_ENABLE_LOGGING"] = "1"

In [4]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 2048 # Can increase for longer reasoning traces
lora_rank = 32 # Larger rank = smarter, but slower

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/Qwen3-4B-Base",
    max_seq_length = max_seq_length,
    load_in_4bit = False, # False for LoRA 16bit
    fast_inference = True, # Enable vLLM fast inference
    max_lora_rank = lora_rank,
    gpu_memory_utilization = 0.7, # Reduce if out of memory
    unsloth_vllm_standby=True
)

model = FastLanguageModel.get_peft_model(
    model,
    r = lora_rank, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = [
        "q_proj", "k_proj", "v_proj", "o_proj",
        "gate_proj", "up_proj", "down_proj",
    ],
    lora_alpha = lora_rank*2, # *2 speeds up training
    use_gradient_checkpointing = "unsloth", # Reduces memory usage
    random_state = 3407,
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.


INFO:unsloth_zoo.temporary_patches.common:Unsloth: Patched Gemma3Processor.__call__.
INFO:unsloth_zoo.temporary_patches.common:Unsloth: Patched AutoHfQuantizer.merge_quantization_configs.
ERROR:unsloth_zoo.temporary_patches.common:Unsloth: Skipped CsmDepthDecoderForCausalLM.forward
Reason: New function removed output_attentions and output_hidden_states
INFO:unsloth_zoo.temporary_patches.common:Unsloth: Patched CsmDepthDecoderForCausalLM.forward.
ERROR:unsloth_zoo.temporary_patches.common:Unsloth: Skipped CsmForConditionalGeneration.forward
Reason: New function removed output_attentions and output_hidden_states
INFO:unsloth_zoo.temporary_patches.common:Unsloth: Patched CsmForConditionalGeneration.forward.
INFO:unsloth_zoo.temporary_patches.common:Unsloth: Patched CsmForConditionalGeneration._merge_input_ids_with_input_values.
INFO:unsloth_zoo.temporary_patches.common:Unsloth: Patched Gemma3nTextAltUp.predict.
INFO:unsloth_zoo.temporary_patches.common:Unsloth: Patched Gemma3nTextAltUp.co

🦥 Unsloth Zoo will now patch everything to make training faster!
INFO 08-12 09:52:41 [importing.py:53] Triton module has been replaced with a placeholder.
INFO 08-12 09:52:41 [__init__.py:239] Automatically detected platform cuda.


INFO:huggingface_hub.file_download:Downloading 'config.json' to '/root/.cache/huggingface/hub/models--unslothai--colab/blobs/c0fd5260818f672a9119328c53cc33df84f30698.incomplete'


Unsloth: Patching vLLM v1 graph capture
Unsloth: Patching vLLM v0 graph capture
==((====))==  Unsloth 2025.8.4: Fast Qwen3 patching. Transformers: 4.55.0. vLLM: 0.8.5.post1.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


Download complete. Moving file to /root/.cache/huggingface/hub/models--unslothai--colab/blobs/c0fd5260818f672a9119328c53cc33df84f30698
INFO:huggingface_hub.file_download:Download complete. Moving file to /root/.cache/huggingface/hub/models--unslothai--colab/blobs/c0fd5260818f672a9119328c53cc33df84f30698
INFO:huggingface_hub.file_download:Downloading 'config.json' to '/root/.cache/huggingface/hub/models--unslothai--colab/blobs/c0fd5260818f672a9119328c53cc33df84f30698.incomplete'
Download complete. Moving file to /root/.cache/huggingface/hub/models--unslothai--colab/blobs/c0fd5260818f672a9119328c53cc33df84f30698
INFO:huggingface_hub.file_download:Download complete. Moving file to /root/.cache/huggingface/hub/models--unslothai--colab/blobs/c0fd5260818f672a9119328c53cc33df84f30698
INFO:huggingface_hub.file_download:Downloading 'config.json' to '/root/.cache/huggingface/hub/models--unslothai--vram-16/blobs/c0fd5260818f672a9119328c53cc33df84f30698.incomplete'
Download complete. Moving file t

Unsloth: vLLM loading unsloth/Qwen3-4B-Base with actual GPU utilization = 69.34%
Unsloth: Your GPU has CUDA compute capability 7.5 with VRAM = 14.74 GB.
Unsloth: Using conservativeness = 1.0. Chunked prefill tokens = 2048. Num Sequences = 160.
Unsloth: vLLM's KV Cache can use up to 3.27 GB. Also swap space = 0 GB.
WARNING 08-12 09:53:01 [config.py:2972] Casting torch.bfloat16 to torch.float16.
INFO 08-12 09:53:20 [config.py:717] This model supports multiple tasks: {'generate', 'reward', 'embed', 'score', 'classify'}. Defaulting to 'generate'.
WARNING 08-12 09:53:20 [arg_utils.py:1658] Compute Capability < 8.0 is not supported by the V1 Engine. Falling back to V0. 
INFO 08-12 09:53:20 [llm_engine.py:240] Initializing a V0 LLM engine (v0.8.5.post1) with config: model='unsloth/Qwen3-4B-Base', speculative_config=None, tokenizer='unsloth/Qwen3-4B-Base', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config=None, tokenizer_revision=None, trust_remote_code=Fals

Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


INFO 08-12 09:54:00 [loader.py:458] Loading weights took 37.97 seconds
INFO 08-12 09:54:00 [punica_selector.py:18] Using PunicaWrapperGPU.
INFO 08-12 09:54:01 [model_runner.py:1140] Model loading took 7.6764 GiB and 38.798665 seconds
INFO 08-12 09:54:11 [worker.py:287] Memory profiling takes 9.37 seconds
INFO 08-12 09:54:11 [worker.py:287] the current vLLM instance can use total_gpu_memory (14.74GiB) x gpu_memory_utilization (0.69) = 10.22GiB
INFO 08-12 09:54:11 [worker.py:287] model weights take 7.68GiB; non_torch_memory takes 0.03GiB; PyTorch activation peak memory takes 0.88GiB; the rest of the memory reserved for KV Cache is 1.64GiB.
INFO 08-12 09:54:11 [executor_base.py:112] # cuda blocks: 745, # CPU blocks: 0
INFO 08-12 09:54:11 [executor_base.py:117] Maximum concurrency for 2048 tokens per request: 5.82x
INFO 08-12 09:54:11 [vllm_utils.py:669] Unsloth: Running patched vLLM v0 `capture_model`.
INFO 08-12 09:54:11 [model_runner.py:1450] Capturing cudagraphs for decoding. This may 

Capturing CUDA graph shapes:   0%|          | 0/23 [00:00<?, ?it/s]

INFO 08-12 09:54:28 [model_runner.py:1592] Graph capturing finished in 17 secs, took 0.34 GiB
INFO 08-12 09:54:28 [vllm_utils.py:676] Unsloth: Patched vLLM v0 graph capture finished in 17 secs.
INFO 08-12 09:54:30 [llm_engine.py:437] init engine (profile, create kv cache, warmup model) took 28.58 seconds
Unsloth: Just some info: will skip parsing ['post_feedforward_layernorm', 'pre_feedforward_layernorm']
Unsloth: Just some info: will skip parsing ['post_feedforward_layernorm', 'pre_feedforward_layernorm']


Unsloth 2025.8.4 patched 36 layers with 36 QKV layers, 36 O layers and 36 MLP layers.


### GRPO chat template
Since we're using a base model, we should set a chat template. You can make your own chat template as well!
1. DeepSeek uses `<think>` and `</think>`, but this is **not** necessary - you can customize it however you like!
2. A `system_prompt` is recommended to at least guide the model's responses.

In [5]:
reasoning_start = "<start_working_out>" # Acts as <think>
reasoning_end   = "<end_working_out>"   # Acts as </think>
solution_start  = "<SOLUTION>"
solution_end    = "</SOLUTION>"

system_prompt = \
f"""You are given a problem.
Think about the problem and provide your working out.
Place it between {reasoning_start} and {reasoning_end}.
Then, provide your solution between {solution_start}{solution_end}"""
system_prompt

'You are given a problem.\nThink about the problem and provide your working out.\nPlace it between <start_working_out> and <end_working_out>.\nThen, provide your solution between <SOLUTION></SOLUTION>'

We create a simple chat template below. Notice `add_generation_prompt` includes prepending `<start_working_out>` to guide the model to start its reasoning process.

In [6]:
chat_template = \
    "{% if messages[0]['role'] == 'system' %}"\
        "{{ messages[0]['content'] + eos_token }}"\
        "{% set loop_messages = messages[1:] %}"\
    "{% else %}"\
        "{{ '{system_prompt}' + eos_token }}"\
        "{% set loop_messages = messages %}"\
    "{% endif %}"\
    "{% for message in loop_messages %}"\
        "{% if message['role'] == 'user' %}"\
            "{{ message['content'] }}"\
        "{% elif message['role'] == 'assistant' %}"\
            "{{ message['content'] + eos_token }}"\
        "{% endif %}"\
    "{% endfor %}"\
    "{% if add_generation_prompt %}{{ '{reasoning_start}' }}"\
    "{% endif %}"

# Replace with out specific template:
chat_template = chat_template\
    .replace("'{system_prompt}'",   f"'{system_prompt}'")\
    .replace("'{reasoning_start}'", f"'{reasoning_start}'")
tokenizer.chat_template = chat_template

Let's see how our chat template behaves on an example:

In [7]:
tokenizer.apply_chat_template([
    {"role" : "user", "content" : "What is 1+1?"},
    {"role" : "assistant", "content" : f"{reasoning_start}I think it's 2.{reasoning_end}{solution_start}2{solution_end}"},
    {"role" : "user", "content" : "What is 2+2?"},
], tokenize = False, add_generation_prompt = True)

"You are given a problem.\nThink about the problem and provide your working out.\nPlace it between <start_working_out> and <end_working_out>.\nThen, provide your solution between <SOLUTION></SOLUTION><|endoftext|>What is 1+1?<start_working_out>I think it's 2.<end_working_out><SOLUTION>2</SOLUTION><|endoftext|>What is 2+2?<start_working_out>"

### Pre fine-tuning for formatting
We now use a subset of NVIDIA's [Open Math Reasoning dataset](https://huggingface.co/datasets/nvidia/OpenMathReasoning) which was filtered to only include high quality DeepSeek R1 traces.

We'll only filter ~59 or so examples to first "prime" / pre fine-tune the model to understand our custom GRPO formatting.

In [8]:
from datasets import load_dataset
import pandas as pd
import numpy as np

dataset = load_dataset("unsloth/OpenMathReasoning-mini", split = "cot")
dataset = dataset.to_pandas()[
    ["expected_answer", "problem", "generated_solution"]
]

# Try converting to number - if not, replace with NaN
is_number = pd.to_numeric(pd.Series(dataset["expected_answer"]), errors = "coerce").notnull()
# Select only numbers
dataset = dataset.iloc[np.where(is_number)[0]]

dataset

expected_answer                                            problem  \
0                  14  Given $\sqrt{x^2+165}-\sqrt{x^2-52}=7$ and $x$...   
6                  -2  Find the value of the parameter $a$ for which ...   
9                  18  What is the sum of all real numbers $x$ for wh...   
13                  2  Evaluate the sum \(\sum_{n=1}^\infty \frac{\ph...   
17                 30  What is the largest positive integer that divi...   
...               ...                                                ...   
19243             244  Let \( p \), \( q \), and \( r \) be the disti...   
19245               1  A bug is on the $0$ of a number line. At any p...   
19247               4  A bus left point X for point Y. Two hours late...   
19248              18  Each interior angle of a regular n-gon measure...   
19250          0.8960  Find the probability that the second blue resu...   

                                      generated_solution  
0      <think>\nOkay, let's see. I need to solve the ...  
6      <think>\nOkay, so I need to find the value of ...  
9      <think>\nOkay, so I need to solve the equation...  
13     <think>\nOkay, so I need to evaluate the infin...  
17     <think>\nAlright, so I need to find the larges...  
...                                                  ...  
19243  <think>\nOkay, so I need to find the value of ...  
19245  <think>\nOkay, so I have this problem where a ...  
19247  <think>\nOkay, let's tackle this problem step ...  
19248  <think>\nOkay, let's see. I need to find the n...  
19250  <think>\nOkay, so I need to find the probabili...  

[7507 rows x 3 columns]

We have to format the dataset to follow our GRPO style formatting:

In [9]:
def format_dataset(x):
    expected_answer = x["expected_answer"]
    problem = x["problem"]

    # Remove generated <think> and </think>
    thoughts = x["generated_solution"]
    thoughts = thoughts.replace("<think>", "").replace("</think>", "")

    # Strip newlines on left and right
    thoughts = thoughts.strip()
    # Add our custom formatting
    final_prompt = \
        reasoning_start + thoughts + reasoning_end + \
        solution_start + expected_answer + solution_end
    return [
        {"role" : "system",    "content" : system_prompt},
        {"role" : "user",      "content" : problem},
        {"role" : "assistant", "content" : final_prompt},
    ]

dataset["Messages"] = dataset.apply(format_dataset, axis = 1)

Check to see if it worked:

In [10]:
tokenizer.apply_chat_template(dataset["Messages"][0], tokenize = False)

"You are given a problem.\nThink about the problem and provide your working out.\nPlace it between <start_working_out> and <end_working_out>.\nThen, provide your solution between <SOLUTION></SOLUTION><|endoftext|>Given $\\sqrt{x^2+165}-\\sqrt{x^2-52}=7$ and $x$ is positive, find all possible values of $x$.<start_working_out>Okay, let's see. I need to solve the equation √(x² + 165) - √(x² - 52) = 7, and find all positive values of x. Hmm, radicals can be tricky, but maybe if I can eliminate the square roots by squaring both sides. Let me try that.\n\nFirst, let me write down the equation again to make sure I have it right:\n\n√(x² + 165) - √(x² - 52) = 7.\n\nOkay, so the idea is to isolate one of the radicals and then square both sides. Let me try moving the second radical to the other side:\n\n√(x² + 165) = 7 + √(x² - 52).\n\nNow, if I square both sides, maybe I can get rid of the square roots. Let's do that:\n\n(√(x² + 165))² = (7 + √(x² - 52))².\n\nSimplifying the left side:\n\nx² + 

Let's truncate the pre fine-tuning dataset to `max_seq_length/2` since we don't want too long reasoning traces.

Note this might take 2 minutes!

In [11]:
dataset["N"] = dataset["Messages"].apply(lambda x: len(tokenizer.apply_chat_template(x)))

dataset = dataset.loc[dataset["N"] <= max_seq_length/2].copy()
dataset.shape

(59, 5)

We then tokenize the messages and convert it to a Hugging Face compatible dataset format:

In [12]:
from datasets import Dataset

dataset["text"] = tokenizer.apply_chat_template(dataset["Messages"].values.tolist(), tokenize = False)
dataset = Dataset.from_pandas(dataset)
dataset

Dataset({
    features: ['expected_answer', 'problem', 'generated_solution', 'Messages', 'N', 'text', '__index_level_0__'],
    num_rows: 59
})

Let's now pre fine-tune the model so it follows our custom GRPO formatting!

In [13]:
from trl import SFTTrainer, SFTConfig
trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    train_dataset = dataset,
    args = SFTConfig(
        dataset_text_field = "text",
        per_device_train_batch_size = 1,
        gradient_accumulation_steps = 1, # Use GA to mimic batch size!
        warmup_steps = 5,
        num_train_epochs = 2, # Set this for 1 full training run.
        learning_rate = 2e-4, # Reduce to 2e-5 for long training runs
        logging_steps = 5,
        optim = "adamw_8bit",
        weight_decay = 0.01,
        lr_scheduler_type = "linear",
        seed = 3407,
        report_to = "none", # Use this for WandB etc
    ),
)

Unsloth: Tokenizing ["text"] (num_proc=2):   0%|          | 0/59 [00:00<?, ? examples/s]

In [14]:
# trainer.train()

Let's check if the model has learnt to follow the custom format:

In [15]:
text = tokenizer.apply_chat_template(
    dataset[0]["Messages"][:2],
    tokenize = False,
    add_generation_prompt = True, # Must add for generation
)

from transformers import TextStreamer
_ = model.generate(
    **tokenizer(text, return_tensors = "pt").to("cuda"),
    temperature = 0,
    max_new_tokens = 1024,
    streamer = TextStreamer(tokenizer, skip_prompt = False),
)

You are given a problem.
Think about the problem and provide your working out.
Place it between <start_working_out> and <end_working_out>.
Then, provide your solution between <SOLUTION></SOLUTION><|endoftext|>Jenifer has 82 cents in pennies and nickels. Her younger brother mistook all her nickels for dimes and counted the total as $1.47. How many pennies does Jenifer have?<start_working_out>Let's denote the number of nickels Jenifer has as N and the number of pennies as P.

We know that each nickel is worth 5 cents and each penny is worth 1 cent. Jenifer has a total of 82 cents, so we can write the following equation:

5N + P = 82

Her younger brother mistook all her nickels for dimes, which are worth 10 cents each. He counted the total as $1.47, which is equivalent to 147 cents. So, we can write another equation:

10N + P = 147

Now we have a system of two equations with two variables:

5N + P = 82
10N + P = 147

We can solve this system by subtracting the first equation from the seco

Yes it did follow the formatting! Great! Let's remove some items before the GRPO step

In [16]:
del dataset
torch.cuda.empty_cache()
import gc
gc.collect()

264

### Data Prep
<a name="Data"></a>

We're using Hugging Face's [Open R1 Math dataset](https://huggingface.co/datasets/open-r1/DAPO-Math-17k-Processed). You can also utilize OpenAI's famous [GSM8K dataset](https://huggingface.co/datasets/openai/gsm8k)

In [17]:
from datasets import load_dataset
dataset = load_dataset("open-r1/DAPO-Math-17k-Processed", "en", split = "train")
dataset

Dataset({
    features: ['prompt', 'solution', 'data_source', 'source_prompt', 'ability', 'reward_model', 'extra_info'],
    num_rows: 14116
})

Let's look at the first row:

In [18]:
dataset[0]["prompt"]

'In triangle $ABC$, $\\sin \\angle A = \\frac{4}{5}$ and $\\angle A < 90^\\circ$. Let $D$ be a point outside triangle $ABC$ such that $\\angle BAD = \\angle DAC$ and $\\angle BDC = 90^\\circ$. Suppose that $AD = 1$ and that $\\frac{BD}{CD} = \\frac{3}{2}$. If $AB + AC$ can be expressed in the form $\\frac{a\\sqrt{b}}{c}$ where $a, b, c$ are pairwise relatively prime integers, find $a + b + c$.'

In [19]:
dataset[0]["solution"]

'34'

In GSM8K, ee notice all answers like about have a ####, so we extract it. But for the Open R1 dataset, we can skip the below.

In [20]:
def extract_hash_answer(text):
    # if "####" not in text: return None
    # return text.split("####")[1].strip()
    return text
extract_hash_answer(dataset[0]["solution"])

'34'

Let's map the dataset! and see the first row:

In [21]:
dataset = dataset.map(lambda x: {
    "prompt" : [
        {"role": "system", "content": system_prompt},
        {"role": "user",   "content": x["prompt"]},
    ],
    "answer": extract_hash_answer(x["solution"]),
})
dataset[0]

{'prompt': [{'content': 'You are given a problem.\nThink about the problem and provide your working out.\nPlace it between <start_working_out> and <end_working_out>.\nThen, provide your solution between <SOLUTION></SOLUTION>',
   'role': 'system'},
  {'content': 'In triangle $ABC$, $\\sin \\angle A = \\frac{4}{5}$ and $\\angle A < 90^\\circ$. Let $D$ be a point outside triangle $ABC$ such that $\\angle BAD = \\angle DAC$ and $\\angle BDC = 90^\\circ$. Suppose that $AD = 1$ and that $\\frac{BD}{CD} = \\frac{3}{2}$. If $AB + AC$ can be expressed in the form $\\frac{a\\sqrt{b}}{c}$ where $a, b, c$ are pairwise relatively prime integers, find $a + b + c$.',
   'role': 'user'}],
 'solution': '34',
 'data_source': 'math_dapo',
 'source_prompt': [{'content': 'Solve the following math problem step by step. The last line of your response should be of the form Answer: $Answer (without quotes) where $Answer is the answer to the problem.\n\nIn triangle $ABC$, $\\sin \\angle A = \\frac{4}{5}$ and $

We create a regex format to match the reasoning sections and answers:

In [22]:
import re

# Add optional EOS token matching
solution_end_regex = r"</SOLUTION>[\s]{0,}" + \
    "(?:" + re.escape(tokenizer.eos_token) + ")?"

match_format = re.compile(
    rf"{reasoning_end}.*?"\
    rf"{solution_start}(.+?){solution_end_regex}"\
    rf"[\s]{{0,}}$",
    flags = re.MULTILINE | re.DOTALL
)
match_format

re.compile(r'<end_working_out>.*?<SOLUTION>(.+?)</SOLUTION>[\s]{0,}(?:<\|endoftext\|>)?[\s]{0,}$',
re.MULTILINE|re.DOTALL|re.UNICODE)

We verify it works:

In [23]:
match_format.findall(
    "Let me think!<end_working_out>"\
    f"<SOLUTION>\n2\n</SOLUTION>",
)

['\n2\n']

In [24]:
match_format.findall(
    "<start_working_out>Let me think!<end_working_out>"\
    f"<SOLUTION>  2  </SOLUTION>\n\n",
)

['  2  ']

We now want to create a reward function to match the format exactly - we reward it with 3 points if it succeeds:

In [25]:
def match_format_exactly(completions, **kwargs):
    scores = []
    for completion in completions:
        score = 0
        response = completion[0]["content"]
        # Match if format is seen exactly!
        if match_format.search(response) is not None: score += 3.0
        scores.append(score)
    return scores

If it fails, we want to reward the model if it at least follows the format partially, by counting each symbol:

In [26]:
def match_format_approximately(completions, **kwargs):
    scores = []
    for completion in completions:
        score = 0
        response = completion[0]["content"]
        # Count how many keywords are seen - we penalize if too many!
        # If we see 1, then plus some points!

        # No need to reward <start_working_out> since we always prepend it!
        # score += 0.5 if response.count(reasoning_start) == 1 else -1.0
        score += 0.5 if response.count(reasoning_end)   == 1 else -1.0
        score += 0.5 if response.count(solution_start)  == 1 else -1.0
        score += 0.5 if response.count(solution_end)    == 1 else -1.0
        scores.append(score)
    return scores

Finally, we want to extract the generated answer, and reward or penalize it! We also reward it based on how close the answer is to the true one via ratios:

In [27]:
def check_answer(prompts, completions, answer, **kwargs):
    question = prompts[0][-1]["content"]
    responses = [completion[0]["content"] for completion in completions]

    extracted_responses = [
        guess.group(1)
        if (guess := match_format.search(r)) is not None else None \
        for r in responses
    ]

    scores = []
    for guess, true_answer in zip(extracted_responses, answer):
        score = 0
        if guess is None:
            scores.append(-2.0)
            continue
        # Correct answer gets 5 points!
        if guess == true_answer:
            score += 5.0
        # Match if spaces are seen, but less reward
        elif guess.strip() == true_answer.strip():
            score += 3.5
        else:
            # We also reward it if the answer is close via ratios!
            # Ie if the answer is within some range, reward it!
            try:
                ratio = float(guess) / float(true_answer)
                if   ratio >= 0.9 and ratio <= 1.1: score += 2.0
                elif ratio >= 0.8 and ratio <= 1.2: score += 1.5
                else: score -= 2.5 # Penalize wrong answers
            except:
                score -= 4.5 # Penalize
        scores.append(score)
    return scores

Also sometimes it might not be 1 number as the answer, but like a sentence for example "The solution is $20" -> we extract 20.

We also remove possible commas for example as in 123,456

In [28]:
match_numbers = re.compile(
    solution_start + r".*?[\s]{0,}([-]?[\d\.\,]{1,})",
    flags = re.MULTILINE | re.DOTALL
)
print(match_numbers.findall("<SOLUTION>  0.34  </SOLUTION>"))
print(match_numbers.findall("<SOLUTION>  123,456  </SOLUTION>"))
print(match_numbers.findall("<SOLUTION>  -0.234  </SOLUTION>"))
print(match_numbers.findall("<SOLUTION>17</SOLUTION>"))

['0.34']
['123,456']
['-0.234']
['17']


We now prepare our main function which will print out the generated responses and the true answer, along with another reward function which converts text to float via `float` and sees if it's the same.

In [29]:
global PRINTED_TIMES
PRINTED_TIMES = 0
global PRINT_EVERY_STEPS
PRINT_EVERY_STEPS = 5

def check_numbers(prompts, completions, answer, **kwargs):
    question = prompts[0][-1]["content"]
    responses = [completion[0]["content"] for completion in completions]

    extracted_responses = [
        guess.group(1)
        if (guess := match_numbers.search(r)) is not None else None \
        for r in responses
    ]

    scores = []
    # Print only every few steps
    global PRINTED_TIMES
    global PRINT_EVERY_STEPS
    if PRINTED_TIMES % PRINT_EVERY_STEPS == 0:
        print(
            '*'*20 + f"Question:\n{question}", f"\nAnswer:\n{answer[0]}", f"\nResponse:\n{responses[0]}", f"\nExtracted:\n{extracted_responses[0]}"
        )
    PRINTED_TIMES += 1

    for guess, true_answer in zip(extracted_responses, answer):
        if guess is None:
            scores.append(-2.5)
            continue
        # Convert to numbers
        try:
            true_answer = float(true_answer.strip())
            # Remove commas like in 123,456
            guess       = float(guess.strip().replace(",", ""))
            scores.append(3.5 if guess == true_answer else -1.5)
        except:
            scores.append(0)
            continue
    return scores

Get the top 90% prompt length so we don't accidentally truncate them!

Ie we'll remove the top 10% long prompts.

In [30]:
tokenized = dataset.map(
    lambda x: {"tokens" : tokenizer.apply_chat_template(x["prompt"], add_generation_prompt = True, tokenize = True)},
    batched = True,
)
print(tokenizer.decode(tokenized[0]["tokens"]))
tokenized = tokenized.map(lambda x: {"L" : len(x["tokens"])})

import numpy as np
maximum_length = int(np.quantile(tokenized["L"], 0.9))
print("Max Length = ", maximum_length)

# Filter only samples smaller than 90% max length
dataset = dataset.select(np.where(np.array(tokenized["L"]) <= maximum_length)[0])
del tokenized

You are given a problem.
Think about the problem and provide your working out.
Place it between <start_working_out> and <end_working_out>.
Then, provide your solution between <SOLUTION></SOLUTION><|endoftext|>In triangle $ABC$, $\sin \angle A = \frac{4}{5}$ and $\angle A < 90^\circ$. Let $D$ be a point outside triangle $ABC$ such that $\angle BAD = \angle DAC$ and $\angle BDC = 90^\circ$. Suppose that $AD = 1$ and that $\frac{BD}{CD} = \frac{3}{2}$. If $AB + AC$ can be expressed in the form $\frac{a\sqrt{b}}{c}$ where $a, b, c$ are pairwise relatively prime integers, find $a + b + c$.<start_working_out>
Max Length =  201


In [31]:
dataset

Dataset({
    features: ['prompt', 'solution', 'data_source', 'source_prompt', 'ability', 'reward_model', 'extra_info', 'answer'],
    num_rows: 12709
})

<a name="Train"></a>
### Train the model

Now set up GRPO Trainer and all configurations!

In [32]:
max_prompt_length = maximum_length + 1 # + 1 just in case!
max_completion_length = max_seq_length - max_prompt_length

from vllm import SamplingParams
vllm_sampling_params = SamplingParams(
    min_p = 0.1,
    top_p = 1.0,
    top_k = -1,
    seed = 3407,
    stop = [tokenizer.eos_token],
    include_stop_str_in_output = True,
)

from trl import GRPOConfig, GRPOTrainer
training_args = GRPOConfig(
    vllm_sampling_params = vllm_sampling_params,
    temperature = 1.0,
    learning_rate = 5e-6,
    weight_decay = 0.01,
    warmup_ratio = 0.1,
    lr_scheduler_type = "linear",
    optim = "adamw_8bit",
    logging_steps = 1,
    per_device_train_batch_size = 1,
    gradient_accumulation_steps = 2, # Increase to 4 for smoother training
    num_generations = 4, # Decrease if out of memory
    max_prompt_length = max_prompt_length,
    max_completion_length = max_completion_length,
    # num_train_epochs = 1, # Set to 1 for a full training run
    max_steps = 100,
    save_steps = 100,
    report_to = "none", # Can use Weights & Biases
    output_dir = "outputs",

    # For optional training + evaluation
    # fp16_full_eval = True,
    # per_device_eval_batch_size = 4,
    # eval_accumulation_steps = 1,
    # eval_strategy = "steps",
    # eval_steps = 1,
)

Unsloth: We now expect `per_device_train_batch_size` to be a multiple of `num_generations`.
We will change the batch size of 1 to the `num_generations` of 4


And let's run the trainer! If you scroll up, you'll see a table of rewards. The goal is to see the `reward` column increase!

You might have to wait 150 to 200 steps for any action. You'll probably get 0 reward for the first 100 steps. Please be patient!

| Step | Training Loss | reward    | reward_std | completion_length | kl       |
|------|---------------|-----------|------------|-------------------|----------|
| 1    | 0.000000      | 0.125000  | 0.000000   | 200.000000        | 0.000000 |
| 2    | 0.000000      | 0.072375  | 0.248112   | 200.000000        | 0.000000 |
| 3    | 0.000000      | -0.079000 | 0.163776   | 182.500000        | 0.000005 |


In [ ]:
# For optional training + evaluation
# new_dataset = dataset.train_test_split(test_size = 0.01)

trainer = GRPOTrainer(
    model = model,
    processing_class = tokenizer,
    reward_funcs = [
        match_format_exactly,
        match_format_approximately,
        check_answer,
        check_numbers,
    ],
    args = training_args,
    train_dataset = dataset,

    # For optional training + evaluation
    # train_dataset = new_dataset["train"],
    # eval_dataset = new_dataset["test"],
)
trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 12,709 | Num Epochs = 1 | Total steps = 100
O^O/ \_/ \    Batch size per device = 4 | Gradient accumulation steps = 2
\        /    Data Parallel GPUs = 1 | Total batch size (4 x 2 x 1) = 8
 "-____-"     Trainable parameters = 66,060,288 of 4,088,528,384 (1.62% trained)
INFO:unsloth_zoo.temporary_patches.common:Unsloth: num_items_in_batch = None


WARNING 08-12 09:57:35 [executor_base.py:215] Executor is not sleeping.
********************Question:
Compute the number of positive integers that divide at least two of the integers in the set $\{1^1,2^2,3^3,4^4,5^5,6^6,7^7,8^8,9^9,10^{10}\}$. 
Answer:
22 
Response:
 
To solve this problem, we need to find the number of positive integers that are divisors of at least two numbers in the set $\{1^1,2^2,3^3,4^4,5^5,6^6,7^7,8^8,9^9,10^{10}\}$. Let's break down the problem step by step:

1. **Prime Factorization:**
   Each number in the set can be expressed in terms of its prime factors:
   - $1^1$ has no prime factors other than 1.
   - $2^2 = 2^2$
   - $3^3 = 3^3$
   - $4^4 = (2^2)^4 = 2^8$
   - $5^5 = 5^5$
   - $6^6 = (2 \cdot 3)^6 = 2^6 \cdot 3^6$
   - $7^7 = 7^7$
   - $8^8 = (2^3)^8 = 2^{24}$
   - $9^9 = (3^2)^9 = 3^{18}$
   - $10^{10} = (2 \cdot 5)^{10} = 2^{10} \cdot 5^{10}$

2. **Identify Common Factors:**
   We need to identify which positive integers can be divisors for at least 

INFO:unsloth_zoo.temporary_patches.common:Unsloth: num_items_in_batch = None


INFO 08-12 09:59:49 [executor_base.py:226] It took 0.640196 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:01:27 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:01:27 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:01:28 [worker.py:145] Sleep mode freed 2.36 GiB memory, 8.88 GiB memory is still in use.
INFO 08-12 10:01:28 [executor_base.py:210] It took 0.642469 seconds to fall asleep.
INFO 08-12 10:01:51 [executor_base.py:226] It took 0.911090 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:01:51 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:01:51 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:01:52 [worker.py:145] Sleep mode freed 1.69 GiB memory, 9.76 GiB memory is still in use.
INFO 08-12 10:01:52 [executor_base.py:210] It took 0.630025 seconds to fall asleep.


INFO:unsloth_zoo.temporary_patches.common:Unsloth: num_items_in_batch = None


INFO 08-12 10:02:10 [executor_base.py:226] It took 0.714442 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:03:34 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:03:34 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:03:34 [worker.py:145] Sleep mode freed 1.85 GiB memory, 8.80 GiB memory is still in use.
INFO 08-12 10:03:34 [executor_base.py:210] It took 0.683286 seconds to fall asleep.
INFO 08-12 10:03:56 [executor_base.py:226] It took 0.847544 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:03:56 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:03:56 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:03:56 [worker.py:145] Sleep mode freed 1.69 GiB memory, 9.50 GiB memory is still in use.
INFO 08-12 10:03:56 [executor_base.py:210] It took 0.679676 seconds to fall asleep.


INFO:unsloth_zoo.temporary_patches.common:Unsloth: num_items_in_batch = None


INFO 08-12 10:04:13 [executor_base.py:226] It took 0.732770 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:04:53 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:04:53 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:04:53 [worker.py:145] Sleep mode freed 1.82 GiB memory, 8.84 GiB memory is still in use.
INFO 08-12 10:04:53 [executor_base.py:210] It took 0.661362 seconds to fall asleep.
INFO 08-12 10:05:07 [executor_base.py:226] It took 0.783884 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:05:07 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:05:07 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:05:08 [worker.py:145] Sleep mode freed 1.69 GiB memory, 9.23 GiB memory is still in use.
INFO 08-12 10:05:08 [executor_base.py:210] It took 0.695577 seconds to fall asleep.


INFO:unsloth_zoo.temporary_patches.common:Unsloth: num_items_in_batch = None


INFO 08-12 10:05:16 [executor_base.py:226] It took 0.901709 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:06:31 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:06:31 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:06:32 [worker.py:145] Sleep mode freed 2.39 GiB memory, 8.81 GiB memory is still in use.
INFO 08-12 10:06:32 [executor_base.py:210] It took 0.759822 seconds to fall asleep.
INFO 08-12 10:06:52 [executor_base.py:226] It took 0.876165 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:06:52 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:06:52 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:06:53 [worker.py:145] Sleep mode freed 1.69 GiB memory, 9.61 GiB memory is still in use.
INFO 08-12 10:06:53 [executor_base.py:210] It took 0.723174 seconds to fall asleep.


INFO:unsloth_zoo.temporary_patches.common:Unsloth: num_items_in_batch = None


INFO 08-12 10:07:07 [executor_base.py:226] It took 0.996918 seconds to wake up tags {'weights', 'kv_cache'}.
********************Question:
Letters $A, B, C,$ and $D$ represent four different digits from 0, 1, 2, 3, ..., 9. If $\frac{A+B}{C+D}$ is an integer that is as large as possible, what is the value of $A+B$? 
Answer:
17 
Response:
1
We are given the equation:

$$
\frac{A+B}{C+D} = \begin{cases}
\text{integer} \\
\text{maximized}
\end{cases}
$$

where $A, B, C,$ and $D$ are four different digits from 0 to 9.

Since $A, B, C,$ and $D$ are different, they must choose digits from 0 to 9 inclusive.

We want to maximize the ratio, so let's start by choosing the largest possible $A$ and $B$. Since the digits are different,

$$
A = 9 \quad \text{and} \quad B = 8
$$

This gives $A+B = 9+8 = 17$. Now, for $C$ and $D$, we want to minimize $C+D$ to maximize the ratio. Therefore, the smaller digits must be chosen for $C$ and $D$. We have 6 remaining digits (0, 1, 2, 3, 4, 5, 7), and we want t

INFO:unsloth_zoo.temporary_patches.common:Unsloth: num_items_in_batch = None


INFO 08-12 10:09:06 [executor_base.py:226] It took 0.843620 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:09:45 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:09:45 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:09:45 [worker.py:145] Sleep mode freed 2.38 GiB memory, 8.82 GiB memory is still in use.
INFO 08-12 10:09:45 [executor_base.py:210] It took 0.795292 seconds to fall asleep.
INFO 08-12 10:09:59 [executor_base.py:226] It took 0.953190 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:09:59 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:09:59 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:10:00 [worker.py:145] Sleep mode freed 1.69 GiB memory, 9.22 GiB memory is still in use.
INFO 08-12 10:10:00 [executor_base.py:210] It took 0.919551 seconds to fall asleep.


INFO:unsloth_zoo.temporary_patches.common:Unsloth: num_items_in_batch = None


INFO 08-12 10:10:08 [executor_base.py:226] It took 0.761752 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:11:41 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:11:41 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:11:42 [worker.py:145] Sleep mode freed 2.40 GiB memory, 8.85 GiB memory is still in use.
INFO 08-12 10:11:42 [executor_base.py:210] It took 0.774203 seconds to fall asleep.
INFO 08-12 10:12:06 [executor_base.py:226] It took 0.974066 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:12:06 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:12:06 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:12:07 [worker.py:145] Sleep mode freed 1.69 GiB memory, 9.56 GiB memory is still in use.
INFO 08-12 10:12:07 [executor_base.py:210] It took 0.723091 seconds to fall asleep.


INFO:unsloth_zoo.temporary_patches.common:Unsloth: num_items_in_batch = None


INFO 08-12 10:12:25 [executor_base.py:226] It took 0.840600 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:13:10 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:13:10 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:13:11 [worker.py:145] Sleep mode freed 2.42 GiB memory, 8.83 GiB memory is still in use.
INFO 08-12 10:13:11 [executor_base.py:210] It took 0.754982 seconds to fall asleep.
INFO 08-12 10:13:26 [executor_base.py:226] It took 0.898638 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:13:26 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:13:26 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:13:27 [worker.py:145] Sleep mode freed 1.69 GiB memory, 9.24 GiB memory is still in use.
INFO 08-12 10:13:27 [executor_base.py:210] It took 0.910741 seconds to fall asleep.


INFO:unsloth_zoo.temporary_patches.common:Unsloth: num_items_in_batch = None


INFO 08-12 10:13:37 [executor_base.py:226] It took 0.815271 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:14:59 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:14:59 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:15:00 [worker.py:145] Sleep mode freed 2.35 GiB memory, 8.85 GiB memory is still in use.
INFO 08-12 10:15:00 [executor_base.py:210] It took 0.795952 seconds to fall asleep.
INFO 08-12 10:15:21 [executor_base.py:226] It took 0.936574 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:15:21 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:15:21 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:15:22 [worker.py:145] Sleep mode freed 1.69 GiB memory, 9.65 GiB memory is still in use.
INFO 08-12 10:15:22 [executor_base.py:210] It took 0.751132 seconds to fall asleep.


INFO:unsloth_zoo.temporary_patches.common:Unsloth: num_items_in_batch = None


INFO 08-12 10:15:38 [executor_base.py:226] It took 0.848234 seconds to wake up tags {'weights', 'kv_cache'}.
********************Question:
How many distinct ordered pairs of positive integers $(m,n)$ are there so that the sum of the reciprocals of $m$ and $n$ is $\frac14$? 
Answer:
5 
Response:
 To find the number of distinct ordered pairs $(m,n)$ of positive integers that satisfy the equation $\frac{1}{m} + \frac{1}{n} = \frac{1}{4}$, we can start by setting a common denominator:

$$ \frac{1}{m} + \frac{1}{n} = \frac{m + n}{mn} = \frac{1}{4}.$$

Then, cross-multiply to get:

$$4(m + n) = mn.$$

Rearranging, we get a quadratic equation in terms of $m$:

$$mn - 4m - 4n = 0.$$

Let's now add 16 to both sides to complete the square:

$$mn - 4m - 4n + 16 = 16.$$

The left-hand side can be factored as:

$$(m - 4)(n - 4) = 16.$$

Given that $m$ and $n$ must be positive integers, the possible pairs $(m - 4, n - 4)$ that multiply to 16 are $(1, 16)$, $(2, 8)$, $(4, 4)$, $(8, 2)$, and $(16, 1)$

INFO:unsloth_zoo.temporary_patches.common:Unsloth: num_items_in_batch = None


INFO 08-12 10:16:55 [executor_base.py:226] It took 1.000364 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:17:49 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:17:49 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:17:50 [worker.py:145] Sleep mode freed 2.35 GiB memory, 8.85 GiB memory is still in use.
INFO 08-12 10:17:50 [executor_base.py:210] It took 0.842019 seconds to fall asleep.
INFO 08-12 10:18:06 [executor_base.py:226] It took 0.913089 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:18:06 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:18:06 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:18:07 [worker.py:145] Sleep mode freed 1.69 GiB memory, 9.27 GiB memory is still in use.
INFO 08-12 10:18:07 [executor_base.py:210] It took 0.778738 seconds to fall asleep.


INFO:unsloth_zoo.temporary_patches.common:Unsloth: num_items_in_batch = None


INFO 08-12 10:18:17 [executor_base.py:226] It took 0.821371 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:19:49 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:19:49 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:19:49 [worker.py:145] Sleep mode freed 1.82 GiB memory, 8.84 GiB memory is still in use.
INFO 08-12 10:19:49 [executor_base.py:210] It took 0.811694 seconds to fall asleep.
INFO 08-12 10:20:13 [executor_base.py:226] It took 1.186272 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:20:13 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:20:13 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:20:14 [worker.py:145] Sleep mode freed 1.69 GiB memory, 9.52 GiB memory is still in use.
INFO 08-12 10:20:14 [executor_base.py:210] It took 0.796172 seconds to fall asleep.


INFO:unsloth_zoo.temporary_patches.common:Unsloth: num_items_in_batch = None


INFO 08-12 10:20:31 [executor_base.py:226] It took 0.864867 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:22:06 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:22:06 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:22:07 [worker.py:145] Sleep mode freed 2.35 GiB memory, 8.85 GiB memory is still in use.
INFO 08-12 10:22:07 [executor_base.py:210] It took 0.876671 seconds to fall asleep.
INFO 08-12 10:22:31 [executor_base.py:226] It took 1.008472 seconds to wake up tags {'weights', 'kv_cache'}.
INFO 08-12 10:22:31 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:22:31 [prefix_caching_block.py:479] Successfully reset prefix cache
INFO 08-12 10:22:32 [worker.py:145] Sleep mode freed 1.69 GiB memory, 9.57 GiB memory is still in use.
INFO 08-12 10:22:32 [executor_base.py:210] It took 0.812379 seconds to fall asleep.


INFO:unsloth_zoo.temporary_patches.common:Unsloth: num_items_in_batch = None


INFO 08-12 10:22:50 [executor_base.py:226] It took 0.828303 seconds to wake up tags {'weights', 'kv_cache'}.


<a name="Inference"></a>
### Inference
Now let's try the model we just trained! First, let's first try the model without any GRPO trained:

In [ ]:
text = "What is the sqrt of 101?"

from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 1.0,
    top_k = 50,
    max_tokens = 1024,
)
output = model.fast_generate(
    [text],
    sampling_params = sampling_params,
    lora_request = None,
)[0].outputs[0].text

output

And now with the LoRA we just trained with GRPO - we first save the LoRA first!

In [ ]:
model.save_lora("grpo_saved_lora")

Verify LoRA is actually trained!

In [ ]:
from safetensors import safe_open

tensors = {}
with safe_open("grpo_saved_lora/adapter_model.safetensors", framework = "pt") as f:
    # Verify both A and B are non zero
    for key in f.keys():
        tensor = f.get_tensor(key)
        n_zeros = (tensor == 0).sum() / tensor.numel()
        assert(n_zeros.item() != tensor.numel())

Now we load the LoRA and test:

In [ ]:
messages = [
    {"role": "system", "content": system_prompt},
    {"role": "user",   "content": "What is the sqrt of 101?"},
]

text = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True, # Must add for generation
    tokenize = False,
)
from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 1.0,
    top_k = 50,
    max_tokens = 2048,
)
output = model.fast_generate(
    text,
    sampling_params = sampling_params,
    lora_request = model.load_lora("grpo_saved_lora"),
)[0].outputs[0].text

output

Our reasoning model is much better - it's not always correct, since we only trained it for an hour or so - it'll be better if we extend the sequence length and train for longer!

<a name="Save"></a>
### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False:
    model.save_pretrained("model")
    tokenizer.save_pretrained("model")
if False:
    model.push_to_hub("hf/model", token = "")
    tokenizer.push_to_hub("hf/model", token = "")


### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

[**NEW**] To finetune and auto export to Ollama, try our [Ollama notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)

In [ ]:
# Save to 8bit Q8_0
if False: model.save_pretrained_gguf("model", tokenizer,)
# Remember to go to https://huggingface.co/settings/tokens for a token!
# And change hf to your username!
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "hf/model", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "",
    )

Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
